kart/annotations/__init__.py

```diff
@@ -0,0 +1,82 @@
+import json
+import logging
+
+from sqlalchemy.exc import OperationalError
+
+from .db import annotations_session, KartAnnotation
+
+L = logging.getLogger(__name__)
+
+
+class DiffAnnotations:
+    def __init__(self, repo):
+        self.repo = repo
+
+    def _object_id(self, base_rs, target_rs):
+        # this is actually symmetric, so we can marginally increase hit rate by sorting first
+        tree_ids = sorted(rs.tree.id for rs in (base_rs, target_rs))
+        return f"{tree_ids[0]}...{tree_ids[1]}"
+
+    def store(self, *, base_rs, target_rs, annotation_type, data):
+        """
+        Stores a diff annotation to the repo's sqlite database,
+        and returns the annotation itself.
+
+        base_rs: base RepoStructure object for this diff (revA in a 'revA...revB' diff)
+        target_rs: target RepoStructure object for this diff (revB in a 'revA...revB' diff)
+        """
+        assert isinstance(data, dict)
+        object_id = self._object_id(base_rs, target_rs)
+        data = json.dumps(data)
+        try:
+            with annotations_session(self.repo) as session:
+                L.debug(
+                    "storing: %s for %s: %s",
+                    annotation_type,
+                    object_id,
+                    data,
+                )
+                session.add(
```

Comment:
presumably this won't create the table if the DB exists but is empty? Do we need a `CREATE TABLE IF NOT EXIST` query on connect?

Reply:
i guess this comes back to the basic behaviour expectations. I assumed:

* if you've run `build-annotations`, populate the DB and use it after that
* if you haven't, don't auto-create the DB.

Does that match your thinking?

Reply:
the table is created and tables populated before this, as soon as we hit `annotations_session` above.

Reply:
That's fine, was just my misunderstanding. Can you add a brief document for this? How it works, when it's created, where it's used, what's stored, etc?

Reply:
Added [ANNOTATIONS.md](https://github.com/koordinates/kart/pull/416/commits/6d32e5db29da54a6744646628905837ec4010fbc)